File: Branches/JSTools 0.41/JSTools.ScriptTypes/JSTools/Util/Serialization/SimpleObjectSerializer.cs
/*
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

using System;
using System.Collections;
using System.Reflection;
using System.Text;

using JSTools.ScriptTypes;

namespace JSTools.Util.Serialization
{
	/// <summary>
	/// Represents an object serializer which serializes/deserialized the
	/// property values and the corresponding property names.
	/// 
	/// The serialized string contains a javascript object, which can
	/// be send to the client browser. To serialize an object hirarchy
	/// you should call the Serialize(...) method. All types which are
	/// derived from IList will be serialized as javascript arrays. All
	/// other complex types will be serialized as javascript objects.
	/// 
	/// To deserialize an object string, you should call the Deserialize(...)
	/// method. All javascript arrays are deserialized into types, which
	/// are derived from IList. If the corresponding type is not derived
	/// from IList, the array values will not be deserilalized. All values
	/// of complex types are serialized into the public properties (requires
	/// a property SETTER). There are two ways to deserialize a javascript
	/// object string:
	/// 
	/// - Deserialize(string, object, bool) method.
	/// You can deserialize the javascript string into an object hirachy.
	/// This is usefull if you have the same object hirachy during
	/// serialization and deserialization. All complex types must be
	/// instantiated in the object hirarchy in order to deserialize all values.
	/// Otherwise some values may be lost. Instances of classes which are
	/// derived from IList and the ReadOnly flag is set to "true" will not be
	/// deserialized.
	/// 
	/// - Deserialize(string, bool) method.
	/// You can deserialize the javascript string into an new object hirachy.
	/// Well, you will obtain JSScriptArray/JSScriptObject objects. This
	/// classes correspond to the javascript array/object types. 
	/// </summary>
	/// <remarks>
	/// Properties which contain a recursion will not be serialized.
	/// </remarks>
	public class SimpleObjectSerializer
	{
		//--------------------------------------------------------------------
		// Declarations
		//--------------------------------------------------------------------

		//--------------------------------------------------------------------
		// Properties
		//--------------------------------------------------------------------

		//--------------------------------------------------------------------
		// Constructors / Destructor
		//--------------------------------------------------------------------

		/// <summary>
		/// Creates a new SimpleObjectSerializer instance.
		/// </summary>
		public SimpleObjectSerializer()
		{
		}

		//--------------------------------------------------------------------
		// Events
		//--------------------------------------------------------------------

		//--------------------------------------------------------------------
		// Methods
		//--------------------------------------------------------------------

		/// <summary>
		/// Deserializes the specified string and returns the deserialized
		/// object structure.
		/// </summary>
		/// <param name="toDeserialize">String to deserialize.</param>
		/// <returns>Returns the created object structure.</returns>
		public object Deserialize(string toDeserialize)
		{
			return Deserialize(toDeserialize, true);
		}

		/// <summary>
		/// Deserializes the specified string and returns the deserialized
		/// object structure.
		/// </summary>
		/// <param name="toDeserialize">String to deserialize.</param>
		/// <param name="decodeValues">True if the values contain characters (e.g. %EF) which should be decoded.</param>
		/// <returns>Returns the created object structure.</returns>
		/// <exception cref="ArgumentNullException">The given string contains a null reference.</exception>
		public object Deserialize(string toDeserialize, bool decodeValues)
		{
			if (toDeserialize == null)
				throw new ArgumentNullException("toDeserialize");

			return new Deserializer().Deserialize(toDeserialize, decodeValues);
		}

		/// <summary>
		/// Deserializes the specified string and fills the deserialized values
		/// into the given object structure.
		/// </summary>
		/// <param name="toDeserialize">String to deserialize.</param>
		/// <param name="toFill">Object structure to fill.</param>
		/// <returns>Returns the object structure.</returns>
		/// <exception cref="ArgumentNullException">The given string contains a null reference.</exception>
		public object Deserialize(string toDeserialize, object toFill)
		{
			return Deserialize(toDeserialize, toFill, true);
		}

		/// <summary>
		/// Deserializes the specified string and fills the deserialized values
		/// into the given object structure.
		/// </summary>
		/// <param name="toDeserialize">String to deserialize.</param>
		/// <param name="toFill">Object structure to fill.</param>
		/// <param name="decodeValues">True if the values contain characters (e.g. %EF) which should be decoded.</param>
		/// <returns>Returns the object structure.</returns>
		/// <exception cref="ArgumentNullException">An argument contains a null reference.</exception>
		public object Deserialize(string toDeserialize, object toFill, bool decodeValues)
		{
			if (toDeserialize == null)
				throw new ArgumentNullException("toDeserialize");

			if (toFill == null)
				throw new ArgumentNullException("toFill");

			return new Deserializer().Deserialize(toDeserialize, toFill, decodeValues);
		}

		/// <summary>
		/// Serializes the given object.
		/// </summary>
		/// <param name="toSerialize">Object which should be serialized.</param>
		/// <returns>Returns the corresponding javascript object string.</returns>
		/// <exception cref="ArgumentNullException">The given object contains a null reference.</exception>
		public string Serialize(object toSerialize)
		{
			return Serialize(toSerialize, false);
		}

		/// <summary>
		/// Serializes the given object.
		/// </summary>
		/// <param name="toSerialize">Object which should be serialized.</param>
		/// <param name="decodeValues">True if the values contain characters (e.g. %) which should be encoded.</param>
		/// <returns>Returns the corresponding javascript object string.</returns>
		/// <exception cref="ArgumentNullException">The given object contains a null reference.</exception>
		public string Serialize(object toSerialize, bool encodeValues)
		{
			if (toSerialize == null)
				throw new ArgumentNullException("toSerialize");

			return new Serializer().Serialize(toSerialize, encodeValues);
		}
	}
}

File: Branches/JSTools 0.50/JSTools.JavaScript.Test/JSTools/Test/SimpleObjectSerializer.js
/*
 * JSTools.JavaScript.Test / JSTools.net - A JavaScript/C# framework.
 * Copyright (C) 2005  Silvan Gehrig
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * Author:
 *  Silvan Gehrig
 */

document.write("<h1>TEST SIMPLEOBJECTSERIALZER</h1>");


var helperRefObj = { 
	aField: 3
}

var objectToSerialize =
{
	__dummy: true,
	a3: 3,
	b6: /ds/gi,
	c0: new Object(),
	d5: new RegExp("[a-z]?", "g"),
	d3i: " \" ||tes t",
	e0$: ' \' \t testy',
	_d3e: false,
	ar: [ objectToSerialize, 38, 0x38, 'string', [ 0, 2, 4 ], new Object(), helperRefObj ],
	ef: helperRefObj
};

// objectToSerialize will not be serialized, it represents a recursion
var serializer = new JSTools.Util.SimpleObjectSerializer();
var serializedObject = serializer.SerializeObject(objectToSerialize);

document.write("<p>serialization output: <br>" + serializedObject + "</p>");

var deserializedObject = serializer.Deserialize(serializedObject);

if (objectToSerialize.a3 != deserializedObject.a3)
	document.write("<p>Deserialization/Serialization failed -> a3 != a3 ('" + objectToSerialize.a3 + "' != '" + deserializedObject.a3 + "')</p>");

if (objectToSerialize.d3i != deserializedObject.d3i)
	document.write("<p>Deserialization/Serialization failed -> d3i != d3i ('" + objectToSerialize.d3i + "' != '" + deserializedObject.d3i + "')</p>");

if (objectToSerialize.e0$ != deserializedObject.e0$)
	document.write("<p>Deserialization/Serialization failed -> e0$ != e0$ ('" + objectToSerialize.e0$ + "' != '" + deserializedObject.e0$ + "')</p>");

if (objectToSerialize.ar[2] != deserializedObject.ar[1])
	document.write("<p>Deserialization/Serialization failed -> ar[2] != ar[1] ('" + objectToSerialize.ar[2] + "' != '" + deserializedObject.ar[1] + "')</p>");

if (objectToSerialize.ar[4][0] != deserializedObject.ar[3][0])
	document.write("<p>Deserialization/Serialization failed -> ar[4][0] != ar[3][0] ('" + objectToSerialize.ar[4][0] + "' != '" + deserializedObject.ar[3][0] + "')</p>");

if (!objectToSerialize.ef)
	document.write("<p>Deserialization/Serialization failed -> ef must contain an instance ('" + ef + "')</p>");

if (deserializedObject.__dummy)
	document.write("<p>Deserialization/Serialization failed -> __dummy must be empty (it's a hidden public field)</p>");

document.write("<p>test done</p>");